Added typing to functions and improved docstrings to be more descriptive.

BeatsuDev opened the pull request:
Added typing to all the functions and parameters in `__init__.py`. Also added Sphinx-styled documentation to explain the parameters of the functions.

Danielhiversen commented:
Thanks.
Please fix the failing tests

BeatsuDev commented:
I think it should be good now 👍

BeatsuDev commented:
I read through the code again and made some small changes/fixes (ref: last two commits on this PR).

Danielhiversen commented:
Great.
Mypy should also be added to test the typing

https://github.com/Danielhiversen/pyTibber/pull/131/files#diff-51335077549f8f95664f409b515ccf2019991e216d97f29b790ddde6c4be4033R46-R48

BeatsuDev commented:
Didn't know about mypy. Seems like a great tool! I'll look into it, thanks! 😄

BeatsuDev commented:
Notable changes after using mypy:
- I changed the declaration of variables from `self.foo: str = None` to `self.foo: str = ""`. This was only changed in the last commit and can be easily reverted if it causes problems.

- Adding to the previous point, I changed the initial value of `self.last_data_timestamp`, `self.peak_hour_time` and `self.last_cons_data_timestamp` to `dt.datetime.min`. I checked the first and last properties for usages and since they are only used to check if they're *lower* than another value, it should be fine.

- Some places I had to add `return None` instead of simply `return` to stop mypy from complaining.

- Since `Tibber.execute` *can* return None, I added an if statement and only set the value of `self.info` if it is defined. I've added TODO comments for potential error handling in the code.

BeatsuDev commented on tibber/__init__.py:
```diff
@@ -480,11 +524,15 @@ async def update_info_and_price_info(self):
+            self.info = data
+            self._process_price_info(self.info)
+        else:
+            # TODO: Error handling?
```
Do you think it's okay to just return None instead of overwriting self.info with a potential None value? Does this break any functionality? There are 2 other places where I've done this (3 in total)

BeatsuDev replied:
@Danielhiversen ^

Danielhiversen replied:
return None is ok for this pr

Danielhiversen commented:
Have you tested your changes locally?

BeatsuDev commented:
> Have you tested your changes locally?

Yep!

BeatsuDev commented:
I actually double-checked and just realised that the tests are being skipped because pytest-asyncio is not installed... The third test then fails because at `assert tibber_connection.name is None`, because the name is an empty string instead (`""`). I'll look into changing all/most the properties into Optional[...] types instead.

BeatsuDev commented:
Tests should run fine now @Danielhiversen. There is one problem that I have chosen to ignore though. 

For some reason, when running pylint, I get this error: `tibber\__init__.py:77:27: E1121: Too many positional arguments for constructor call (too-many-function-args)`. The workflow in GitHub does not catch this however, and the error doesn't make sense anyway since the amount of arguments are correct... So I've chosen to ignore it.

You should also take a look at line 311.

BeatsuDev commented on tibber/__init__.py:
```diff
@@ -269,7 +307,9 @@ async def fetch_consumption_data(self):
             else:
-                n_hours = (now - self.last_cons_data_timestamp).total_seconds() / 3600
+                n_hours = int(
+                    (now - self.last_cons_data_timestamp).total_seconds() / 3600 # type: ignore[operator]
```
This is the line where I ignore a mypy error. `self.last_cons_data_timestamp` will always be defined at this point, right?

Danielhiversen commented on requirements.txt:
```diff
@@ -0,0 +1,50 @@
+# All the requirements needed for the project + running tests
```
Please do not add any more changes to this pr.
So no need to add requirements.txt

Danielhiversen commented on .github/workflows/code_checker.yml:
```diff
@@ -20,24 +20,25 @@ jobs:
+          cache: pip
+      - run: pip install -r requirements.txt
 
-      - uses: actions/cache@v2
```
Why disable the cache?

BeatsuDev commented:
I added a new dependency (pytest-asyncio) so that the tests could be run and not just be skipped (see issue #178). There were some issues with the cache so the setup of the workflow failed. Therefore removed the caching step and used python-setup's built-in caching feature instead. This required a requirements.txt to install from afaik, so I added that.

BeatsuDev commented:
I could revert these last changes (requirements.txt and pytest-asyncio) and add them in a different PR if you'd like. Shall I do that?

Danielhiversen commented on .github/workflows/code_checker.yml:
```diff
@@ -29,12 +29,19 @@ jobs:
       - name: Pylint Code Linter
         run: |
             pylint tibber
+
```
```suggestion
```

Danielhiversen commented on .github/workflows/code_checker.yml:
```diff
@@ -29,12 +29,18 @@ jobs:
       - name: isort
         run: |
             isort **/*.py
+
```
```suggestion
```

Danielhiversen commented on requirements.txt:
```diff
@@ -1 +1 @@
-.
+.
```
```suggestion
.

```

BeatsuDev commented:
Not sure why pylint did not use the setup.cfg in this branch, but it worked fine in the master branch... Anyway, I fixed it by shortening the line 👍
